flat_mutation_reader.cc

```diff
@@ -1344,6 +1345,13 @@ flat_mutation_reader downgrade_to_v1(flat_mutation_reader_v2 r) {
         }
         virtual future<> next_partition() override {
             clear_buffer_to_next_partition();
+
+            // If there is an active tombstone at the moment of doing next_partition(), simply clearing the buffer
+            // means that the closing range_tombstone_change will not be processed by the assembler. This violates
+            // an assumption of the assembler. Since the client executed next_partition(), they aren't interested
+            // in currently active tombstone_change (if any). Therefore in both cases, next partition should start
+            // with the assembler state reset.
+            _rt_assembler.reset();
```

Comment:
I think it's wrong to reset the assembler here. _rt_assembler is part of the producer's state, and producer may be ahead of the consumer. For example, the producer may be inside partition P2 now, and the consumer which calls next_partition() may be inside P1, skipping into P2. In such case, resetting _rt_assembler will corrupt the assembler.

We should reset it only if `is_buffer_empty()`.

Reply:
Agreed. I made the requested change.